File: 3366-minimum-array-sum/3366-minimum-array-sum.java
class Solution {
    public int minArraySum(int[] nums, int k, int op1, int op2) {
        int n = nums.length;
        int[][][] dp = new int[n + 1][op1 + 1][op2 + 1];

        for(int[][] row : dp) {
            for(int[] col : row) {
                Arrays.fill(col, -1);
            }
        }

        
        return f(nums, k, op1, op2, 0, dp);
    }

    private int f(int[] nums, int k, int op1, int op2, int index,int[][][] dp) {
        if(index >= nums.length) {
            return 0;
        }

        if(dp[index][op1][op2] != -1) return dp[index][op1][op2];

        int ele = nums[index];

        int sum = ele + f(nums, k, op1, op2, index + 1, dp);
        if(op1 > 0) {
            int val = (nums[index] + 1) / 2;
            sum = Math.min(sum, val + f(nums, k, op1 -1, op2, index + 1, dp));
        }

        if(op2 > 0 & nums[index] - k >= 0) {
            int val = nums[index] - k;
            sum = Math.min(sum, val + f(nums, k, op1, op2 -1 , index + 1, dp));
        }

        if(op1 > 0 && op2 > 0 && nums[index] - k >= 0) {
            int val =(nums[index] + 1) / 2;
            if(val - k >= 0){
                val = val - k;
                sum = Math.min(sum, val + f(nums, k, op1 -1, op2 -1, index + 1, dp));
            }
            val =  (nums[index] - k + 1) / 2;
            sum = Math.min(sum, val + f(nums, k, op1 -1, op2 -1, index + 1, dp));
            
        }

        return dp[index][op1][op2] = sum;
        
    }
}

File: 3366-minimum-array-sum/3366-minimum-array-sum.py
class Solution:
    def minArraySum(self, nums: List[int], k: int, op1: int, op2: int) -> int:
        n = len(nums)
        dp = [[[0 for k in range(n+1)] for i in range(n+1)] for j in range(n+1)]
        for i in range(1,n+1):
            for o1 in range(op1+1):
                for o2 in range(op2+1):
                    minn = dp[i-1][o1][o2] + nums[i-1]
                    if o1>0:
                        minn = min(
                            minn,
                            dp[i-1][o1-1][o2] + ceil(nums[i-1]/2)
                        )
                    if o2>0 and nums[i-1] >= k:
                        minn = min(
                            minn, 
                            dp[i-1][o1][o2-1]+nums[i-1]-k,
                        )
                    if o1>0 and o2>0 and ceil(nums[i-1]/2)>=k:
                        minn = min(
                            minn, 
                            dp[i-1][o1-1][o2-1]+ceil(nums[i-1]/2)-k
                        )
                    if o1>0 and o2>0 and nums[i-1]>=k:
                        minn = min(
                            minn, 
                            dp[i-1][o1-1][o2-1]+ceil((nums[i-1]-k)/2)
                        )

                    dp[i][o1][o2] = minn
        return dp[n][op1][op2]